src/manifold/src/smoothing.cpp

```diff
@@ -308,6 +191,312 @@ struct InterpTri {
+ public:
+  // The cached partitions don't have idx - it's added to the copy returned
+  // from GetPartition that contains the mapping of the input divisions into the
+  // sorted divisions that are uniquely cached.
+  glm::ivec3 idx;
+  glm::ivec3 sortedDivisions;
+  Vec<glm::vec3> vertBary;
+  Vec<glm::ivec3> triVert;
+
+  int InteriorOffset() const {
+    return sortedDivisions[0] + sortedDivisions[1] + sortedDivisions[2];
+  }
+
+  int NumInterior() const { return vertBary.size() - InteriorOffset(); }
+
+  static Partition GetPartition(glm::ivec3 divisions) {
+    glm::ivec3 sortedDiv = divisions;
+    glm::ivec3 triIdx = {0, 1, 2};
+    if (sortedDiv[2] > sortedDiv[1]) {
+      std::swap(sortedDiv[2], sortedDiv[1]);
+      std::swap(triIdx[2], triIdx[1]);
+    }
+    if (sortedDiv[1] > sortedDiv[0]) {
+      std::swap(sortedDiv[1], sortedDiv[0]);
+      std::swap(triIdx[1], triIdx[0]);
+      if (sortedDiv[2] > sortedDiv[1]) {
+        std::swap(sortedDiv[2], sortedDiv[1]);
+        std::swap(triIdx[2], triIdx[1]);
+      }
+    }
+
+    Partition partition = GetCachedPartition(sortedDiv);
+    partition.idx = triIdx;
+
+    return partition;
+  }
+
+  Vec<glm::ivec3> Reindex(glm::ivec3 tri, glm::ivec3 edgeOffsets,
+                          glm::bvec3 edgeFwd, int interiorOffset) const {
+    Vec<int> newVerts;
+    newVerts.reserve(vertBary.size());
+    glm::ivec3 triIdx = idx;
+    std::vector<int> outTri = {0, 1, 2};
+    if (idx[1] != Next3(idx[0])) {
+      triIdx = {idx[2], idx[0], idx[1]};
+      edgeFwd = glm::not_(edgeFwd);
+      std::swap(outTri[0], outTri[1]);
+    }
+    for (const int i : {0, 1, 2}) {
+      newVerts.push_back(tri[triIdx[i]]);
+    }
+    for (const int i : {0, 1, 2}) {
+      const int n = sortedDivisions[i] - 1;
+      int offset = edgeOffsets[idx[i]] + (edgeFwd[idx[i]] ? 0 : n - 1);
+      for (int j = 0; j < n; ++j) {
+        newVerts.push_back(offset);
+        offset += edgeFwd[idx[i]] ? 1 : -1;
+      }
+    }
+    const int offset = interiorOffset - newVerts.size();
+    for (int i = newVerts.size(); i < vertBary.size(); ++i) {
+      newVerts.push_back(i + offset);
+    }
+
+    Vec<glm::ivec3> newTriVert(triVert.size());
+    for (int i = 0; i < triVert.size(); ++i) {
+      for (const int j : {0, 1, 2}) {
+        newTriVert[i][outTri[j]] = newVerts[triVert[i][j]];
+      }
+    }
+    return newTriVert;
+  }
+
+ private:
+  static inline auto cacheLock = std::mutex();
+  static inline auto cache =
+      std::unordered_map<glm::ivec3, std::unique_ptr<Partition>>();
+
+  // This triangulation is purely topological - it depends only on the number of
+  // divisions of the three sides of the triangle. This allows them to be cached
+  // and reused for similar triangles. The shape of the final surface is defined
+  // by the tangents and the barycentric coordinates of the new verts. The input
+  // must be sorted: n[0] >= n[1] >= n[2] > 0
+  static Partition GetCachedPartition(glm::ivec3 n) {
```

Comment:
Maybe we will want a limit on the size of the cache, and make it a LRU or something. Otherwise there may be memory leak in long running applications. Perhaps we can define the limit in `ManifoldParams`?

Reply:
Let's leave that for a follow-on PR. I think it's pretty unlikely the cache will be that huge, and if it is, it means it's being used.